### PR TITLE
ElasticSearch reporter

### LINKED ISSUE #10: [Elasticsearch] ES Reporter Implementation
https://github.com/alhardy/AppMetrics/issues/94

### PR DESCRIPTION
**Pull request by jenyayel**
Few notes:

- I noticed you did some progress in branch 1.0.0. LMK how you want to handle it.
- I took great inspiration (steal) from the InfluxDB reported you had. The implementation is very similar, which probably will be ease for you to understand. On another hand, there is some common code base which can be shared between those (or may be future) reporters.
- There are some internal methods (e.g. `String` extensions) that npt visible to the new assembly, so I duplicated them. They all under `App.Metrics.Internal` namespace.
- I didn't write tests, since heavily relied on method of InfluxDB. LMK how you think will be best handle shared code, and where you think tests should be added.
- Each measurement is stored in separate ES type. This is similar to how it was done in original Metrics.NET library. LMK if you want to change it.
- All metrics are presented under object `fields`, the document structure is as following:
```
{
   "name": "",
   "fields": { },
   "tags": { },
   "written": "2017-03-19T12:18:25.4036296Z"
}
```
So the `timer` looks like this:
```
{
   "name": "application.httprequests__http_requests",
   "fields": {
	  "countMeter": 2068,
	  "rate1m": 871.7556391420713,
	  "rate5m": 727.2766300771132,
	  "rate15m": 670.6556474002623,
	  "rateMean": 851.1950174738226,
	  "samples": 1028,
	  "last": 351.33351,
	  "countHist": 2068,
	  "min": 6.84946,
	  "max": 10885.21464,
	  "mean": 1350.3788217804288,
	  "median": 1245.0691399999998,
	  "stddev": 1047.8925081784437,
	  "p999": 8830.27619,
	  "p99": 3129.73746,
	  "p98": 3054.25769,
	  "p95": 2948.50393,
	  "p75": 2277.7939499999998
   },
   "tags": {
	  "env": "stage",
	  "app": "sample app"
   },
   "written": "2017-03-19T12:18:25.4026303Z"
}
```

and the `counter`:
```
{
   "name": "appmetrics.internal__report_success",
   "fields": {
	  "value": 1
   },
   "tags": {
	  "env": "stage",
	  "app": "sample app"
   },
   "written": "2017-03-19T12:18:25.4036296Z"
}
```

If everything above is acceptable:
- I already updated Samples project and will open PR
- I'll update DocFx in upcoming days
- Kibana examples I hope to add right after it

Closes #10, alhardy/AppMetrics#94

### CODE REVIEW

**Comment by jenyayel** [comment]
I noticed an issue that `fields.value` may contain in case of `health` type textual value and in case of `counter `  integer - not ideal situation in case you want to aggregate without defining mappings in ES. Do you have an idea may be how to model it differently?

**Comment by alhardy** [comment]
@jenyayel 😎 

I'm going to be releasing a v1 shortly which unfortunately may break some functionality here. I'll get back to you once that's released and work through the changes with your guidance on the ES side of things.

Thanks for this PR!

**Comment by alhardy** [comment]
@jenyayel 

>I noticed an issue that fields.value may contain in case of health type textual value and in case of counter integer - not ideal situation in case you want to aggregate without defining mappings in ES. Do you have an idea may be how to model it differently?

This has been changed in 1.0.0, health check names are reported as tags and are provided as a metric, see [here](https://github.com/alhardy/AppMetrics/blob/1.0.0/src/App.Metrics/Health/Internal/DefaultHealthProvider.cs#L58). Therefore reporters don't need to handle health check reporting in any special way

**Comment by jenyayel** [comment]
> I'm going to be releasing a v1 shortly which unfortunately may break some functionality here. I'll get back to you once that's released and work through the changes with your guidance on the ES side of things.

That is totally fine. I did see the progress you made which makes a lot of sense. I'm not in rush, so LMK when you want me to rebase/merge.

**Comment by alhardy** [comment]
@jenyayel merged [1.0.0](https://github.com/alhardy/AppMetrics/releases/tag/1.0.0) to master

**Comment by jenyayel** [comment]
@alhardy merged and migrated.

I also removed health reporting. 

As I mentioned (in examples above) all measurements reported under "generic" nested object `fields`. I picked this direction because this is how it was implemented in Metrics.NET. If you take a look at original *codahale metrics* [ES sink](https://github.com/elastic/elasticsearch-metrics-reporter-java/blob/master/README.md#timer), they placed each measurement in dedicated field. This makes document slightly nicer. What do you think?

File: src/App.Metrics.Extensions.Reporting.ElasticSearch/Extensions/MetricValueExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+            values.Add("samples", apdex.SampleSize);
+            values.Add("score", apdex.Score);
+            values.Add("satisfied", apdex.Satisfied);
+            values.Add("tolerating", apdex.Tolerating);
+            values.Add("frustrating", apdex.Frustrating);
+        }
+
+        public static void AddHistogramValues(this HistogramValue histogram, IDictionary<string, object> values)
+        {
+            values.Add("samples", histogram.SampleSize);
```
Comment (alhardy) [inline src/App.Metrics.Extensions.Reporting.ElasticSearch/Extensions/MetricValueExtensions.cs]: Histograms now include a 'sum' see https://github.com/alhardy/AppMetrics.Reporters/issues/6#issuecomment-280363290.

Just noticed I missed this in the influxdb reporter as well ;)

Comment (jenyayel) [inline src/App.Metrics.Extensions.Reporting.ElasticSearch/Extensions/MetricValueExtensions.cs]: Added `sum`. I also refactored those extensions to mach Influx - another common area.

**Comment by jenyayel** [comment]
> It'd be great if you could try that out on a prod application an let me know how it goes. 

Cannot promise this, but sure I'll push it to our staging and will give feedback. 

Thanks for the project ;)

**Comment by alhardy** [comment]
@jenyayel Haha, staging is safer yes ;)

I'm using the InfluxDB reporter on a few prod api's, been fine so far, but yes looking forward to your feedback.

You'll have to give me until after the weekend to publish your ES reporter package.

> Thanks for the project ;)

No worries, thanks for contributing!

**Pull request merged**